### PR TITLE
Rollerbar intake

### PR DESCRIPTION
**Pull request by J-Double-J**
This creates 2 commands that run rollerbar in and out. It sets gamepad and wiring configurations, and adds a new subsystem (RollerbarIntakeSubsystem) to robotSubsystems that is set up. RollerbarIntake uses a double solenoid.

### CODE REVIEW

**Comment by mabster314** [comment]
I will take a look at this as soon as I can -- probably sometime tomorrow.

File: src/org/usfirst/frc2175/command/single/RunRollerbarIntakeInCommand.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    private RollerbarIntakeSubsystem rollerbarIntakeSubsystem;
+
+    public RunRollerbarIntakeInCommand(RobotSubsystems robotSubsystems,
+            double speed) {
+        rollerbarIntakeSubsystem =
+                robotSubsystems.getRollerbarIntakeSubsystem();
+        this.speed = speed;
+
+        requires(rollerbarIntakeSubsystem);
+
```
Comment (bvisness) [inline src/org/usfirst/frc2175/command/single/RunRollerbarIntakeInCommand.java]: Might as well get rid of this extra line of whitespace. (Line 23)

File: src/org/usfirst/frc2175/config/GamepadConfig.java [hunk 2]
```diff
@@ -30,10 +32,16 @@ public String getPropertyFileName() {
...
-
-        testAction = new JoystickButton(gamepad, getIntPropertyValue("button.action.test", properties));
+        extendCatapult = new JoystickButton(gamepad,
+                getIntPropertyValue("button.catapult.extend", properties));
+        retractCatapult = new JoystickButton(gamepad,
+                getIntPropertyValue("button.catapult.retract", properties));
+        runIntakeIn = new JoystickButton(gamepad,
+                getIntPropertyValue("button.catapult.intake.in", properties));
+        runIntakeOut = new JoystickButton(gamepad,
+                getIntPropertyValue("button.catapult.intake.out", properties));
```
Comment (bvisness) [inline src/org/usfirst/frc2175/config/GamepadConfig.java]: The intake is not really part of the catapult, so a better property name might just be `button.intake.out`. Same for `.in` above. And don't forget to change this in gamepad.properties.

File: src/properties/practicebot/gamepad.properties [hunk 3]
```diff
@@ -1,3 +1,8 @@
 gamepad.port = 2
+
 button.catapult.extend = 1
 button.catapult.retract = 2
+button.catapult.intake.in = 3
+button.catapult.intake.out = 4
+
+button.action.test = 3
```
Comment (bvisness) [inline src/properties/practicebot/gamepad.properties]: Let's put `button.action.test` on a button that's less common to press. I would suggest 10.

Basically, I just don't want to forget that we have testing on a common button. I don't think there's anything to prevent multiple commands from being mapped to the same button by accident.

File: src/org/usfirst/frc2175/commandmapper/JoystickEventMapper.java [hunk 4]
```diff
@@ -27,7 +29,16 @@ public JoystickEventMapper(RobotConfig robotConfig,
         retractCatapult
                 .whenPressed(new RetractCatapultCommand(robotSubsystems));
 
+        JoystickButton runIntakeIn = gamepadConfig.getIntakeInButton();
+        runIntakeIn.whenPressed(
+                new RunRollerbarIntakeInCommand(robotSubsystems, 0));
+
+        JoystickButton runIntakeOut = gamepadConfig.getIntakeOutButton();
+        runIntakeOut.whenPressed(
+                new RunRollerbarIntakeOutCommand(robotSubsystems, 0));
```
Comment (bvisness) [inline src/org/usfirst/frc2175/commandmapper/JoystickEventMapper.java]: These don't make sense, because according to how the commands are set up these will run the intake with a speed of 0. But take a look at my overall comment on this PR.

**Comment by bvisness** [comment]
Structurally, this looks fantastic! Everything is where it is supposed to be and I can't think of anything you missed. So :+1: to that!

But...let's think about the commands themselves a bit. You created two commands, `RunRollerbarIntakeInCommand` and `RunRollerbarIntakeOutCommand`. But both of them require you to tell the command what speed to run the roller bar at. The "out" command just negates the speed so it goes backwards.

I would prefer that we do what we did for the dream. Namely:
1. Have one command in the `single` package called `RunRollerbarIntakeAtSpeedCommand`. This would work exactly like the current `RunRollerbarIntakeInCommand`.
2. Modify `IntakeConfig` to load the default in and out speeds from a properties file.
3. Modify the `RunIntakeInGroup` and `RunIntakeOutGroup` to use the rollerbar intake instead of the dream intake. (And, of course, load the appropriate values from `IntakeConfig`.)

Great job on this overall though! Once those commands are in order I think we can merge this.

File: src/org/usfirst/frc2175/subsystem/intake/RollerbarIntakeSubsystem.java [hunk 5]
```diff
@@ -1,7 +1,48 @@
...
+
+    public void setRollerbarSpeed(double speed) {
+        rollerbarIntakeTalon.set(speed);
+    }
+
+    public void setIntakePosition(boolean isOut) {
+        if (isOut) {
+            rollerbarIntakeSolenoid.set(DoubleSolenoid.Value.kReverse);
+        } else {
+            rollerbarIntakeSolenoid.set(DoubleSolenoid.Value.kForward);
```
Comment (bvisness) [inline src/org/usfirst/frc2175/subsystem/intake/RollerbarIntakeSubsystem.java:35]: Is this backwards? Currently if `isOut` is true, the solenoid will be set to the reverse position.

File: src/org/usfirst/frc2175/subsystem/intake/RollerbarIntakeSubsystem.java [hunk 6]
```diff
@@ -1,7 +1,48 @@
...
+        }
+    }
+
+    public boolean isCompletelyOut() {
+        return rollerbarIntakeDownSwitch.get();
+    }
 
+    public boolean isCompletelyIn() {
+        return rollerbarIntakeUpSwitch.get();
+    }
```
Comment (bvisness) [inline src/org/usfirst/frc2175/subsystem/intake/RollerbarIntakeSubsystem.java:47]: Do we refer to the intake's position as down/up or in/out? We need to decide on this to reduce confusion.

File: src/org/usfirst/frc2175/Robot.java [hunk 7]
```diff
@@ -22,9 +23,10 @@
             new DriverStation(robotConfig, new DeadbandCalculator());
     private final RobotSubsystems robotSubsystems =
             new RobotSubsystems(robotConfig, driverStation);
+    private final IntakeConfig intakeConfig = new IntakeConfig();
     private final JoystickEventMapper joystickEventMapper =
-            new JoystickEventMapper(robotConfig, driverStation,
-                    robotSubsystems);
+            new JoystickEventMapper(robotConfig, driverStation, robotSubsystems,
+                    intakeConfig);
```
Comment (bvisness) [inline src/org/usfirst/frc2175/Robot.java]: `intakeConfig` should be included in `robotConfig`, so this shouldn't be necessary.

File: src/org/usfirst/frc2175/command/group/RunIntakeInGroup.java [hunk 8]
```diff
@@ -1,26 +1,19 @@
...
-        double sideBeltSpeed = intakeConfig.getSideBeltSpeedForward();
-        double mainBeltSpeed = intakeConfig.getMainBeltSpeedForward();
+        double rollerbarSpeed = intakeConfig.getRollerbarSpeedIn();
 
-        addParallel(new RunDreamIntakeSideBeltAtSpeedCommand(robotSubsystems,
-                sideBeltSpeed));
-        addSequential(new RunDreamIntakeMainBeltAtSpeedCommand(robotSubsystems,
-                mainBeltSpeed));
+        addSequential(new RunRollerbarIntakeAtSpeedCommand(robotSubsystems,
+                rollerbarSpeed));
```
Comment (bvisness) [inline src/org/usfirst/frc2175/command/group/RunIntakeInGroup.java:16]: :+1:

File: src/org/usfirst/frc2175/command/group/RunIntakeOutGroup.java [hunk 9]
```diff
@@ -1,26 +1,19 @@
...
-        double sideBeltSpeed = intakeConfig.getSideBeltSpeedReverse();
-        double mainBeltSpeed = intakeConfig.getMainBeltSpeedReverse();
+        double rollerbarSpeed = intakeConfig.getRollerbarSpeedOut();
 
-        addParallel(new RunDreamIntakeSideBeltAtSpeedCommand(robotSubsystems,
-                sideBeltSpeed));
-        addSequential(new RunDreamIntakeMainBeltAtSpeedCommand(robotSubsystems,
-                mainBeltSpeed));
+        addSequential(new RunRollerbarIntakeAtSpeedCommand(robotSubsystems,
+                rollerbarSpeed));
```
Comment (bvisness) [inline src/org/usfirst/frc2175/command/group/RunIntakeOutGroup.java:17]: Looks like there's some unnecessary whitespace at the bottom of these constructors. You might as well get rid of it.

File: src/org/usfirst/frc2175/command/single/RunRollerbarIntakeAtSpeedCommand.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ */
+public class RunRollerbarIntakeAtSpeedCommand extends Command {
+
+    private double speed;
+    private RollerbarIntakeSubsystem rollerbarIntakeSubsystem;
+
+    public RunRollerbarIntakeAtSpeedCommand(RobotSubsystems robotSubsystems,
+            double speed) {
+        rollerbarIntakeSubsystem =
```
Comment (bvisness) [inline src/org/usfirst/frc2175/command/single/RunRollerbarIntakeAtSpeedCommand.java]: Let's do `this.rollerbarIntakeSubsystem` here for consistency.

File: src/org/usfirst/frc2175/command/single/RunRollerbarIntakeAtSpeedCommand.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+                robotSubsystems.getRollerbarIntakeSubsystem();
+        this.speed = speed;
+
+        requires(rollerbarIntakeSubsystem);
```
Comment (bvisness) [inline src/org/usfirst/frc2175/command/single/RunRollerbarIntakeAtSpeedCommand.java]: And here too.

**Comment by bvisness** [comment]
We don't need `RunRollerbarIntakeInCommand` or `RunRollerbarIntakeOutCommand` any more, since we have `RunRollerbarIntakeAtSpeedCommand`.

File: src/org/usfirst/frc2175/commandmapper/JoystickEventMapper.java [hunk 12]
```diff
@@ -16,7 +19,8 @@
  */
 public class JoystickEventMapper {
     public JoystickEventMapper(RobotConfig robotConfig,
-            DriverStation driverStation, RobotSubsystems robotSubsystems) {
+            DriverStation driverStation, RobotSubsystems robotSubsystems,
+            IntakeConfig intakeConfig) {
```
Comment (bvisness) [inline src/org/usfirst/frc2175/commandmapper/JoystickEventMapper.java]: We don't need `intakeConfig` here, since it's part of `robotConfig`...

File: src/org/usfirst/frc2175/commandmapper/JoystickEventMapper.java [hunk 13]
```diff
@@ -16,7 +19,8 @@
...
         GamepadConfig gamepadConfig = robotConfig.getGamepadConfig();
```
Comment (bvisness) [inline src/org/usfirst/frc2175/commandmapper/JoystickEventMapper.java:23]: So you should pull it out of `robotConfig` here, like with `gamepadConfig`.

File: src/org/usfirst/frc2175/commandmapper/JoystickEventMapper.java [hunk 14]
```diff
@@ -27,7 +31,16 @@ public JoystickEventMapper(RobotConfig robotConfig,
         retractCatapult
                 .whenPressed(new RetractCatapultCommand(robotSubsystems));
 
+        JoystickButton runIntakeIn = gamepadConfig.getIntakeInButton();
+        runIntakeIn.whenPressed(new RunRollerbarIntakeInCommand(robotSubsystems,
+                intakeConfig.getRollerbarSpeedIn()));
```
Comment (bvisness) [inline src/org/usfirst/frc2175/commandmapper/JoystickEventMapper.java]: Use `RunIntakeInGroup` here instead of `RunRollerbarIntakeInCommand`. Same below for running it outward.

File: src/org/usfirst/frc2175/commandmapper/JoystickEventMapper.java [hunk 15]
```diff
@@ -27,7 +31,16 @@ public JoystickEventMapper(RobotConfig robotConfig,
...
+
+        JoystickButton runIntakeOut = gamepadConfig.getIntakeOutButton();
+        runIntakeOut.whenPressed(new RunRollerbarIntakeOutCommand(
+                robotSubsystems, intakeConfig.getRollerbarSpeedOut()));
+
         JoystickButton testAction = gamepadConfig.getTestActionButton();
         testAction.whenPressed(new TurnToFaceGoalCommand(robotSubsystems));
+
```
Comment (bvisness) [inline src/org/usfirst/frc2175/commandmapper/JoystickEventMapper.java]: No need for this line of whitespace.

File: src/org/usfirst/frc2175/subsystem/RobotSubsystems.java [hunk 16]
```diff
@@ -22,6 +24,7 @@ public RobotSubsystems(RobotConfig robotConfig,
         this.driverStation = driverStation;
 
         dreamIntakeSubsystem = new DreamIntakeSubsystem(robotConfig);
```
Comment (bvisness) [inline src/org/usfirst/frc2175/subsystem/RobotSubsystems.java]: We can remove `dreamIntakeSubsystem` from `RobotSubsystems` by the sound of things. Don't delete any dream-related files though.

File: src/properties/practicebot/wiring.properties [hunk 17]
```diff
@@ -17,3 +21,5 @@ catapult.solenoid.left.forward = 0
 catapult.solenoid.left.reverse = 1
 catapult.solenoid.right.forward = 2
 catapult.solenoid.right.reverse = 3
+rollerbarIntake.solenoid.forward = 4
+rollerbarIntake.solenoid.reverse = 5
```
Comment (bvisness) [inline src/properties/practicebot/wiring.properties]: Missed the extra line at the end of the file here.

File: src/properties/competitionbot/intake.properties [hunk 18]
```diff
@@ -2,3 +2,5 @@ dream.intake.sideBelt.forward = 1
 dream.intake.sideBelt.reverse = -1
 dream.intake.mainBelt.forward = 1
 dream.intake.mainBelt.reverse = -1
+rollerbar.intake.in = -1
+rollerbar.intake.out = 1
```
Comment (bvisness) [inline src/properties/competitionbot/intake.properties]: I would suggest naming these property `rollerbar.intakeSpeed.in` and `rollerbar.intakeSpeed.out`. Just calling it "intake" doesn't really communicate what the property means.

Don't forget to change this in `IntakeConfig` as well.

File: src/org/usfirst/frc2175/commandmapper/JoystickEventMapper.java [hunk 19]
```diff
@@ -27,6 +29,14 @@ public JoystickEventMapper(RobotConfig robotConfig,
         retractCatapult
                 .whenPressed(new RetractCatapultCommand(robotSubsystems));
 
+        JoystickButton runIntakeIn = gamepadConfig.getIntakeInButton();
+        runIntakeIn.whenPressed(new RunIntakeInGroup(robotSubsystems,
+                robotConfig.getIntakeConfig()));
```
Comment (bvisness) [inline src/org/usfirst/frc2175/commandmapper/JoystickEventMapper.java]: Rather than doing `robotConfig.getIntakeConfig` here, pull intakeConfig out to its own variable up above. Like on line 22 of this file!

Like this: `IntakeConfig intakeConfig = robotConfig.getIntakeConfig();`

You can put that on line 23.

Comment (bvisness) [inline src/org/usfirst/frc2175/commandmapper/JoystickEventMapper.java]: I should say that yes, your code does work! It's just that we can clean it up a little.

Comment (bvisness) [inline src/org/usfirst/frc2175/commandmapper/JoystickEventMapper.java]: You still missed this one... :smile:

**Comment by bvisness** [comment]
I think it's finally good! I'm not gonna bother to squash all the commits together this time, so you can click the big green Merge button!
